metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,362 @@
+    (finder, decoder)
+  }
+
+  private def getTastyDecoder(
+      params: Map[String, String]
+  ): (Finder, Decoder) = {
+    val finder = Finder(uriPath =>
+      for {
+        path <- toFile(uriPath, ".tasty-detailed")
+        classesPath <-
+          if (path.isScalaOrJava) findClassesDirFileFromSource(path, "tasty")
+          else findPathInfoFromClassesPath(path)
+      } yield classesPath
+    )
+    val decoder = Decoder(decodeFromTastyFile(_))
+    (finder, decoder)
+  }
+
+  private def findPathInfoFromClassesPath(
+      path: AbsolutePath
+  ): Option[PathInfo] = {
+    val pathInfos = for {
+      scalaTarget <- buildTargets.all
+      classPath = scalaTarget.classDirectory.toAbsolutePath
+      if (path.isInside(classPath))
+    } yield PathInfo(Some(scalaTarget.id), path)
+    pathInfos.toList.headOption
+  }
+
+  private def findClassesDirFileFromSource(
+      sourceFile: AbsolutePath,
+      newExtension: String
+  ): Option[PathInfo] = {
+    for {
+      targetId <- buildTargets.sourceBuildTargets(sourceFile).headOption
+      target <- buildTargets.scalaTarget(targetId)
+      sourceRoot <- buildTargets.inverseSourceItem(sourceFile)
+      classDir = target.classDirectory.toAbsolutePath
+      oldExtension = sourceFile.extension
+      relativePath = sourceFile
+        .toRelative(sourceRoot)
+        .resolveSibling(_.stripSuffix(oldExtension) + newExtension)
+    } yield PathInfo(Some(targetId), classDir.resolve(relativePath))
+  }
+  private def findSemanticDBFileFromSource(
+      sourceFile: AbsolutePath
+  ): Option[PathInfo] = {
+    for {
+      targetId <- buildTargets.inverseSources(sourceFile)
+      target <- buildTargets.scalaTarget(targetId)
+      sourceRoot <- buildTargets.workspaceDirectory(targetId)
+      targetRoot = target.targetroot
+      relativePath = SemanticdbClasspath.fromScala(
+        sourceFile.toRelative(sourceRoot.dealias)
+      )
+      foundSemanticDbPath <- fileSystemSemanticdbs.findSemanticDb(
+        relativePath,
+        targetRoot,
+        sourceFile,
+        workspace
+      )
+    } yield PathInfo(Some(targetId), foundSemanticDbPath.path)
+  }
+
+  private def decodeJavapFromClassFile(
+      pathInfo: PathInfo,
+      verbose: Boolean
+  ): Future[Option[String]] = {
+    try {
+      val args = if (verbose) List("-verbose") else Nil
+      val sb = new StringBuilder()
+      shellRunner
+        .run(
+          "Decode using javap",
+          JavaBinary(userConfig().javaHome, "javap") :: args ::: List(
+            pathInfo.path.filename
+          ),
+          pathInfo.path.parent,
+          redirectErrorOutput = true,
+          Map.empty,
+          s => {
+            sb.append(s)
+            sb.append(Properties.lineSeparator)
+          },
+          s => (),
+          propagateError = true,
+          logInfo = false
+        )
+        .map(_ => Some(sb.toString))
+    } catch {
+      case _: Throwable => Future.successful(None)
```

Comment:
More Scala way to catch exceptions (because of fatal errors and `ControlTrowable`)
```suggestion
      case NonFatal(_) => Future.successful(None)
```